File: validation/qiskit/deutsch_jozsa.py
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector


def constant_zero(n: int) -> QuantumCircuit:
    return QuantumCircuit(n + 1)


def constant_one(n: int) -> QuantumCircuit:
    qc = QuantumCircuit(n + 1)
    answer = n
    qc.x(answer)
    return qc


def balanced_selected_bit(n: int, query: int) -> QuantumCircuit:
    qc = QuantumCircuit(n + 1)
    answer = n
    qc.cx(query, answer)
    return qc


def balanced_parity(n: int) -> QuantumCircuit:
    qc = QuantumCircuit(n + 1)
    answer = n

    for query in range(n):
        qc.cx(query, answer)

    return qc


def deutsch_jozsa_circuit(n: int, oracle: QuantumCircuit) -> QuantumCircuit:
    answer = n
    qc = QuantumCircuit(n + 1)
    qc.x(answer)
    qc.h(range(n + 1))
    qc.compose(oracle, inplace=True)
    qc.h(range(n))

    return qc


def query_zero_probability(state: Statevector, n: int) -> float:
    probs = state.probabilities()
    query_mask = (1 << n) - 1

    p_zero = 0.0
    for basis_index, probability in enumerate(probs):
        query_bits = basis_index & query_mask
        if query_bits == 0:
            p_zero += probability

    return p_zero


def classify(p_zero: float) -> str:
    return "constant" if p_zero > 0.5 else "balanced"


def run_case(label: str, expected: str, n: int, oracle: QuantumCircuit) -> None:
    qc = deutsch_jozsa_circuit(n, oracle)
    state = Statevector.from_instruction(qc)

    p_zero = query_zero_probability(state, n)
    observed = classify(p_zero)

    assert observed == expected, (
        f"{label}: expected {expected}, got {observed}, "
        f"P(query = |00...0>) = {p_zero:.6f}"
    )

    print(f"{label}:")
    print(f"  expected: {expected}")
    print(f"  observed: {observed}")
    print(f"  P(query = |00...0>): {p_zero:.6f}")
    print()


def main() -> None:
    n = 3

    cases = [
        ("constant zero", "constant", constant_zero(n)),
        ("constant one", "constant", constant_one(n)),
        ("balanced selected bit", "balanced", balanced_selected_bit(n, query=0)),
        ("balanced parity", "balanced", balanced_parity(n)),
    ]

    for label, expected, oracle in cases:
        run_case(label, expected, n, oracle)


if __name__ == "__main__":
    main()

File: src/algorithms/deutsch_jozsa.rs
use crate::algorithms::deutsch_helpers::{
    ANSWER, classify_deutsch_jozsa_result, query_register_zero_probability,
};
use crate::circuit::core::Circuit;
use crate::engine::utils::to_dirac;
use crate::examples::catalog::deutsch_jozsa_circuit;
use ndarray::ArrayD;
use num_complex::Complex64;
use std::ops::RangeInclusive;
// Deutsch-Jozsa algorithm demo.
//
// This module implements the Deutsch-Jozsa algorithm for Boolean oracles that
// are guaranteed to be either constant or balanced. The circuit uses one answer
// qubit and one or more query qubits.
//
// Qubit layout:
// - ANSWER is qubit 0
// - query qubits are 1..=num_query_qubits
//
// The oracle applies Uf |x, y> = |x, y xor f(x)>.
// After the final Hadamard's on the query register, the result is classified by
// checking the probability of the query register being |00...0>.
//
// Expected behavior:
// - constant oracle -> query register returns |00...0>
// - balanced oracle -> query register does not return |00...0>

#[derive(Debug, Clone, Copy)]
pub enum DeutschJozsaOracle {
    ConstantZero,
    ConstantOne,
    BalancedSingleBit { query: usize },
    BalancedParity,
}
pub struct DeutschJozsaResult {
    pub expected_class: &'static str,
    pub observed_class: &'static str,
    pub p_query_zero: f64,
    pub state: ArrayD<Complex64>,
}

impl DeutschJozsaOracle {
    pub(crate) fn apply(self, circuit: &mut Circuit, num_query_qubits: usize) {
        match self {
            DeutschJozsaOracle::ConstantZero => {}
            DeutschJozsaOracle::ConstantOne => {
                circuit.x(ANSWER);
            }
            DeutschJozsaOracle::BalancedSingleBit { query } => {
                assert!(
                    (1..=num_query_qubits).contains(&query),
                    "query must be in 1..={}",
                    num_query_qubits
                );
                circuit.cnot(query, ANSWER);
            }
            DeutschJozsaOracle::BalancedParity => {
                for query in query_qubits(num_query_qubits) {
                    circuit.cnot(query, ANSWER);
                }
            }
        }
    }
}

impl DeutschJozsaOracle {
    fn label(self) -> &'static str {
        match self {
            DeutschJozsaOracle::ConstantZero => "f(x) = 0",
            DeutschJozsaOracle::ConstantOne => "f(x) = 1",
            DeutschJozsaOracle::BalancedSingleBit { .. } => "f(x) = selected query bit",
            DeutschJozsaOracle::BalancedParity => "f(x) = parity of query register",
        }
    }
    fn expected_class(self) -> &'static str {
        match self {
            DeutschJozsaOracle::ConstantZero | DeutschJozsaOracle::ConstantOne => "constant",
            DeutschJozsaOracle::BalancedSingleBit { .. } | DeutschJozsaOracle::BalancedParity => {
                "balanced"
            }
        }
    }
}

fn query_qubits(num_query_qubits: usize) -> RangeInclusive<usize> {
    1..=num_query_qubits
}

pub(crate) fn apply_h_to_queries(circuit: &mut Circuit, num_query_qubits: usize) {
    for query in query_qubits(num_query_qubits) {
        circuit.h(query);
    }
}

fn deutsch_jozsa_state(num_query_qubits: usize, oracle: DeutschJozsaOracle) -> ArrayD<Complex64> {
    deutsch_jozsa_circuit(num_query_qubits, oracle).run()
}

pub fn run_deutsch_jozsa(
    num_query_qubits: usize,
    oracle: DeutschJozsaOracle,
) -> DeutschJozsaResult {
    let state = deutsch_jozsa_state(num_query_qubits, oracle);

    let p_query_zero = query_register_zero_probability(&state, num_query_qubits);
    let observed_class = classify_deutsch_jozsa_result(&state, num_query_qubits);

    DeutschJozsaResult {
        expected_class: oracle.expected_class(),
        observed_class,
        p_query_zero,
        state,
    }
}

pub fn run_deutsch_jozsa_demo() {
    let num_query_qubits = 3;

    let oracles = [
        DeutschJozsaOracle::ConstantZero,
        DeutschJozsaOracle::ConstantOne,
        DeutschJozsaOracle::BalancedSingleBit { query: 1 },
        DeutschJozsaOracle::BalancedParity,
    ];

    println!("\nDeutsch-Jozsa algorithm:\n");

    for oracle in oracles {
        let result = run_deutsch_jozsa(num_query_qubits, oracle);

        println!("   Oracle: {}", oracle.label());
        println!("   Query qubits: {}", num_query_qubits);
        println!("   Expected class: {}", result.expected_class);
        println!("   Observed class: {}", result.observed_class);
        println!(
            "   P(query register = |00...0>): {:.6}",
            result.p_query_zero
        );
        println!("   Final state:");
        println!("   {}", to_dirac(&result.state));
        println!()
    }
}
